### PR TITLE
feat: Created installation script for mentorship-backend

### LINKED ISSUE #223: Create an Installation script to automatically install and setup the project
**Is your feature request related to a problem? Please describe.**
Currently the installation and setup of Mentorship Backend is done manually. It would be great to have a single script which could download the code and setup the various configurations required automatically. 

**Describe the solution you'd like**
Create a shell script named **install.sh** (for Linux/Mac) or **install.bat** (for Windows) which would do the following when run - 
    - Clone the repository to the present working directory of the running script (if not already inside one, which will mostly be the case )
    - Perform all the required setup for the repository to run. Take any variable inputs as mandatory command line arguments, without which the script will produce an error, and prompt the user to enter the values. 
    - Perform unit testing on the installation and display its results

**Describe alternatives you've considered**
To modify the deploy.sh file to work on local systems

### PR DESCRIPTION
**Pull request by rudrakshjhalanii07**
### Description
Created an installation script for mentorship-backend which clones the project repo, sets up the environment variables, tests the file and run it.

Fixes #223 

### Type of Change:
**Delete irrelevant options.**

- Code
- Quality Assurance
- Documentation

**Code/Quality Assurance Only**
- This change requires a documentation update (software upgrade on readme file)
- New feature (non-breaking change which adds functionality pre-approved by mentors)

### How Has This Been Tested?
Has been tested by me manually.

### Checklist:
**Delete irrelevant options.**

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] Any dependent changes have been merged
- [ ] Update Postman API at /docs folder
- [ ] Update Swagger documentation and the exported file at /docs folder
- [ ] Update requirements.txt

**Code/Quality Assurance Only**
- [x] My changes generate no new warnings 
- [ ] My PR currently breaks something (fix or feature that would cause existing functionality to not work as expected)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been published in downstream modules

### CODE REVIEW

File: install.bat
Comment (xprilion) [inline install.bat]: I'm not sure if this works this way... 

`source` is a UNIX command. The `env/Scripts/activate` is for Windows

File: install.bat
Comment (xprilion) [inline install.bat]: Does this work on Windows? `export` is a UNIX command.

**Review by xprilion — CHANGES_REQUESTED** [review]
At the end of the installation script, it is useful to display the name of the environment which was created automatically or else the user will not know how to run it after the installation is done. Also, suggesting helpful links pointing to the Wiki at any place where the setup fails would be great!

**Comment by rudrakshjhalanii07** [comment]
I ran the script in bash so it worked there, my bad.
I shall quickly fix it up and will report to you.

**Review by xprilion — CHANGES_REQUESTED** [review]
Hi @Hardik0403, I tried running this script on Windows 10 CMD and got these errors - 

```
C:\Users\Anubhav\Desktop\mentorship-backend>install.bat

C:\Users\Anubhav\Desktop\mentorship-backend>#Cloning the git repository
'#Cloning' is not recognized as an internal or external command,
operable program or batch file.

C:\Users\Anubhav\Desktop\mentorship-backend>#git clone https://github.com/systers/mentorship-backend.git
'#git' is not recognized as an internal or external command,
operable program or batch file.

C:\Users\Anubhav\Desktop\mentorship-backend>#This method confirms whether user has entered an input or not
'#This' is not recognized as an internal or external command,
operable program or batch file.

C:\Users\Anubhav\Desktop\mentorship-backend>valid_input(){
'valid_input' is not recognized as an internal or external command,
operable program or batch file.

C:\Users\Anubhav\Desktop\mentorship-backend>message="$1"
'message' is not recognized as an internal or external command,
operable program or batch file.

C:\Users\Anubhav\Desktop\mentorship-backend>config_val="$2"
'config_val' is not recognized as an internal or external command,
operable program or batch file.

C:\Users\Anubhav\Desktop\mentorship-backend>while [ true ]
'while' is not recognized as an internal or external command,
operable program or batch file.

C:\Users\Anubhav\Desktop\mentorship-backend>do
'do' is not recognized as an internal or external command,
operable program or batch file.
-z was unexpected at this time.

C:\Users\Anubhav\Desktop\mentorship-backend>if [ -z $config_val ]

C:\Users\Anubhav\Desktop\mentorship-backend>
```

Can you please check on them? Maybe I did something wrong? Do let me know!

**Comment by rudrakshjhalanii07** [comment]
So sorry for my mistakes. I've updated the file although!

File: install.bat
Comment (xprilion) [inline install.bat]: Can you please make this optional? Maybe asking the user before the entire process begins would be great since he/she would usually already have a clone to be able to have this file, but sometimes he/she may want a clean setup!

File: install.bat
Comment (xprilion) [inline install.bat]: Can you also please show the environment name and how to activate it after the setup finishes?

File: install.bat
Comment (xprilion) [inline install.bat:59]: If the tests fail, we should not get a success message as would the next set of lines display! maybe you want to check for errors here

File: install.bat
Comment (xprilion) [inline install.bat:24]: This step may fail at times, we need to display an error if it does

**Review by xprilion — CHANGES_REQUESTED** [review]
Thanks for the changes @Hardik0403, it does work now but I have a few concerns! Kindly go through the comments in the review

**Comment by rudrakshjhalanii07** [comment]
I have carefully observed the asked changes and have rectified them.

File: install.bat
Comment (BadduCoder) [inline install.bat]: Give it a python version (python3) which it will use for creating virtual environment. 
Problem arises when python2 is used by default, so explicitly mention the usage of python3.

Comment (Hardik0403) [reply install.bat]: Sorry but I couldn't understand it clearly, 
Do you mean to say that I shall take the name of python version from user as a input, like by which name the user has his python path mentioned?

File: install.bat
Comment (sumit-badsara) [inline install.bat:28]: I guess you need to put exit command here as the setup fails?

File: install.bat
Comment (sumit-badsara) [inline install.bat:17]: Give it a python version (python3) which it will use for creating virtual environment.
Problem arises when python2 is used by default, so explicitly mention the usage of python3.

**Review by sumit-badsara — COMMENTED** [review]
Just some minor corrections needed :smiley: !

File: install.bat
Comment (rudrakshjhalanii07) [reply install.bat:17]: I have made this correct!

File: install.bat
Comment (rudrakshjhalanii07) [reply install.bat:28]: I have transferred the control to the : steps function so that the user gets to know how to activate virtual env. Again if he wants to, the program would be exited from there automatically.

Comment (sumit-badsara) [reply install.bat:28]: Okay @Hardik0403! That would be fine too I guess. As @xprilion was reviwing your task, let him get back to the comments he left, once they are approved! I guess your task is complete. Great work :+1:

Comment (rudrakshjhalanii07) [reply install.bat:28]: Thank you for your feedback!
Yes sure, I would be waiting for the review.☺️

Comment (sumit-badsara) [reply install.bat:28]: In case it takes him more than expected time, some other mentor will approve the changes and finalize the task. Thanks for the patience. :smile:

Comment (sumit-badsara) [reply install.bat:28]: @Hardik0403 Also, you can submit it for the review on code-in now. As your part is done, next response is needed from mentor.

Comment (rudrakshjhalanii07) [reply install.bat:28]: No issues!! I remember he told that there is an Internet issue going on in his region, so I would be waiting.☺️

File: install.bat
Comment (sumit-badsara) [inline install.bat]: This is a minor error that I came across now, everything else is fine. 
It should be
IF %check_repo%==y
without double quotes as it directly compares the input.

**Review by sumit-badsara — CHANGES_REQUESTED** [review]
@Hardik0403 I will finalize the task today as @xprilion is having trouble with his internet connection. 
Just a minor correction and we are good to go.

File: install.bat
Comment (rudrakshjhalanii07) [reply install.bat]: Done with the changes.

**Review by sumit-badsara — DISMISSED** [review]
This works great! All done. Great work @Hardik0403 💯

**Comment by rudrakshjhalanii07** [comment]
Thank you so much!😊

File: install.bat
Comment (xprilion) [inline install.bat]: Hey @Hardik0403, this should not be executed if we're not cloning the repo. You would want to move it inside the if-else clause above!

File: install.bat
Comment (xprilion) [inline install.bat:52]: This value can be calculated (which will be the safer option) by taking a substring of the `MAIL_DEFAULT_SENDER`. Can you try using that?

**Review by xprilion — CHANGES_REQUESTED** [review]
Hi @Hardik0403, great work, the script looks much more refined now! 

Kindly go through my comments, one of which is critical for the script to function. 

Also, there's an important UX point to consider here - 
when you're working with a language/software you like to know what was the error, why did it happen and how to solve it. While you've mentioned all errors that happen well, by messages like - 

```
ECHO "Error while testing the file. Exited with status : %errorlevel%"
```
It might be great if you could add a link to somewhere in the wiki or an authentic source of help when the error happens, right below the error message. 

The above is just a nice-to-have, as @BadduCoder has already approved your task it would not count explicitly towards your contributions in GCI but definitely help with the community rep :slightly_smiling_face:

File: install.bat
Comment (rudrakshjhalanii07) [reply install.bat]: I assumed that install.bat file is outside the mentorship-backend folder with the user, I have now fixed this.

**Comment by rudrakshjhalanii07** [comment]
I have updated all the changes, please let me know if there are more places for improvement and then I shall squash the commits into one!
Thank you @xprilion @BadduCoder for your advices!

**Review by robotjellyzone — CHANGES_REQUESTED** [review]
Hi @Hardik0403 well done 👍  !! please, can you squash your commits into one so, that it will be easier for us to merge that updated PR only . To know more about this, you can search through the github's guide .

**Review by robotjellyzone — DISMISSED** [review]
I am approving this PR, since you will find difficulty in squashing so, that part will also be done by us during merging this PR.
PR ready to merge 👍

**Comment by vj-codes** [comment]
Closing since issue is not assigned to the author 
Thank you for your contribution :)

**Pull request closed without merging**
